sync/src/block/downloader/body.rs

```diff
@@ -16,101 +16,142 @@
     }
 
-    pub fn remove_target(&mut self, targets: &[BlockHash]) {
+    pub fn remove_targets(&mut self, targets: &[BlockHash]) {
         if targets.is_empty() {
             return
         }
         cdebug!(SYNC, "Remove download targets: {:?}", targets);
-        for hash in targets {
-            if let Some(index) = self.targets.iter().position(|t| t == hash) {
-                self.targets.remove(index);
-            }
-            self.downloading.remove(hash);
-            self.downloaded.remove(hash);
-        }
+        // XXX: It can be slow.
+        self.states.retain(|hash, _| !targets.contains(hash));
+        self.targets.retain(|hash| !targets.contains(hash));
+        self.states.shrink_to_fit();
         self.targets.shrink_to_fit();
-        self.downloading.shrink_to_fit();
-        self.downloaded.shrink_to_fit();
     }
 
     pub fn reset_downloading(&mut self, hashes: &[BlockHash]) {
         cdebug!(SYNC, "Remove downloading by timeout {:?}", hashes);
         for hash in hashes {
-            self.downloading.remove(&hash);
+            if let Some(state) = self.states.get_mut(hash) {
+                if *state == State::Downloading {
+                    *state = State::Queued;
+                }
+            }
         }
-        self.downloading.shrink_to_fit();
     }
 
     pub fn drain(&mut self) -> Vec<(BlockHash, Vec<UnverifiedTransaction>)> {
         let mut result = Vec::new();
-        for t in &self.targets {
-            if let Some(body) = self.downloaded.remove(t) {
-                result.push((*t, body));
-            } else {
-                break
+        for hash in &self.targets {
+            let entry = self.states.entry(*hash);
+            let state = match entry {
+                Entry::Vacant(_) => unreachable!(),
+                Entry::Occupied(mut entry) => match entry.get_mut() {
+                    state @ State::Downloaded {
+                        ..
+                    } => replace(state, State::Drained),
+                    _ => break,
+                },
+            };
+            match state {
+                State::Downloaded {
+                    transactions,
+                } => {
+                    result.push((*hash, transactions));
+                }
+                _ => unreachable!(),
             }
         }
-        self.downloaded.shrink_to_fit();
-        self.targets.drain(0..result.len());
-        self.targets.shrink_to_fit();
         result
     }
 
     pub fn re_request(&mut self, hash: BlockHash, remains: Vec<(BlockHash, Vec<UnverifiedTransaction>)>) {
+        let hashes = remains.iter().map(|(hash, _)| *hash);
         let mut new_targets = vec![hash];
-        new_targets.extend(remains.into_iter().map(|(hash, transactions)| {
-            self.downloaded.insert(hash, transactions);
-            hash
-        }));
+        new_targets.extend(hashes);
         new_targets.append(&mut self.targets);
         self.targets = new_targets;
```

Comment:
Do we still add block hashes to the `self.target` here?
The reason why the previous `re_request` added block hashes to the `self.target` is that the `drain` function was removing block_hashes from the `self.target`.

Since the new `drain` function is not removing block hashes from the `self.target`. We must not add them to the `self.target` here.

Reply:
You are right. It must not add the targets.
I'll fix it.